mathics/builtin/assignments/internals.py

```diff
@@ -567,9 +567,12 @@ def process_assign_makeboxes(self, lhs, rhs, evaluation, tags, upset):
     #   MakeBoxes[CubeRoot, StandardForm] := RadicalBox[3, StandardForm]
     # rather than:
     #   MakeBoxes[CubeRoot, StandardForm] -> RadicalBox[3, StandardForm]
-    makeboxes_rule = Rule(lhs, rhs, system=True)
-    makeboxes_defs = evaluation.definitions.builtin["System`MakeBoxes"]
-    makeboxes_defs.add_rule(makeboxes_rule)
+
+    makeboxes_rule = Rule(lhs, rhs, system=False)
+    definitions = evaluation.definitions
+    definitions.add_rule("System`MakeBoxes", makeboxes_rule, "down")
+    #    makeboxes_defs = evaluation.definitions.builtin["System`MakeBoxes"]
```

Comment:
During a session, we should not touch the `builtin` definition of a symbol, but ask to the `Definitions` object to add the new rule in the proper place. In general, the proper place is in an `user` definition, which is then merged with the `builtin` definition when it is needed.

Reply:
Ok. Thanks for the explanation and fix of what is wrong.

Why is is there a definition inside the symbol at all? My gut feeling is that if this is removed, then Symbol is more like a Lisp Symbol.  But to start out with, let me narrow this question for just SymbolMakeBoxes symbols. 

The definitions we are adding _are_  (or will be) the first ones added, so what is the value of having _any_ rule associated with SymbolMakeBoxes?

Reply:
I am not sure to understand your question. A `Symbol` in WL is a literal `Atom`, just like a `String` or an `Integer`. With all of them, you can build `Expression` s,  which are just a list of Atoms. 

Then we have Definitions, that we can associate to Symbols. We could have more than one Definition for the same Symbol. A Definition stores the attributes of a Symbol, together with a set (or several sets) of replacement rules. 





Then, we have certain actions we can perform over these elements (Atoms and Expressions). Replacements are one of these the possible actions. In a replacement, we take an element, and we produce a new element following certain rules, and eventually, producing certain side actions on the system (like write a file, run a system command). 

Another kind of actions that we can perform over an Expression is to "evaluate" it. For doing it,  






> Ok. Thanks for the explanation and fix of what is wrong.
> 
> Why is is there a definition inside the symbol at all? My gut feeling is that if this is removed, then Symbol is more like a Lisp Symbol (and we start to see more speed improvement ;-). But to start out with, let me narrow this question for just SymbolMakeBoxes symbols.
> 

OK. I am not sure to understand the question. I will try again to provide a description of the elements, and probably you can reformulate them in more standard/generic/LISP-compliant terms.

Symbols do not have a Definition "inside". A `Symbol` is one kind of the `Atom` (ic) `Elements` over which `Expression`s are made. `Elements` and `Expressions` are the basic data type of the WL.
 `Elements` have several methods that requires a set of definitions as arguments, but if we strip them, I think we could have a minimal "lexical" module which do not need to make any references to `Definitions` or `Evaluation` classes, where the state of a (`Mathics`) `Session` is stored.

Then, we can associate a `Definition` with a given `Symbol`. A `Definition` stores certain attributes and replacement rules associated to a `Symbol` and to the expression built from these symbols. Then, an `Evaluation` (*evaluation context*? /  *session context* / *command line session*) is defined by a set of Definitions (a `Definitions` object) and other internal registers (timeouts,  semaphores, expected formats, character encoding, etc, which for some reason we do not want to store in the form of Definitions).


Over an `Expression`, we can perform different actions that produce a new `Expression`. One possible action could be to sort/ rearrange the elements of an expression, or to replace a `Symbol` with another `Element`. 
 Another example would be to *apply* a `Rule`. In this kind of action, a part of an Expression that matches a certain pattern is replaced by another expression that can depend on the matching expression. If the rule is successfully applied, then we get a new Expression. In the process of building the new expression, the old expression could have side effects, like running a system command, writting a file,  or changing the state of the session.

To `evaluate` an `Expression` consists then in take it, and applying interactively and recursively the replacement rules associated to the `Symbol`s appearing in the `Expression` in a certain order, to produce a new `Expression`. To help picking the right/pertinent rules during the evaluation, each Definition has different sets of rules (`Ownvalues`, `Downvalues`, `Upvalues`, `NValues`,  `Format`, `Default`, etc.). 

If we can think about a unique evaluation loop, certain symbols lead to specific evaluation patterns, involving following a different order in the application of the rules. One example are the expressions of the form  `N[...]`.   These expressions look at a different set of rules (`NValues`) instead of the one used in a regular evaluation (`Ownvalues`,  `Downvalues`,  `Upvalues`, etc).  `Format` and `MakeBoxes` does something similar.


> The definitions we are adding _are_ (or will be) the first ones added, so what is the value of having _any_ rule associated with SymbolMakeBoxes?

In the current implementation, the evaluation of `MakeBoxes[...]` follows the same mechanism that the used for any other expression. And these kind of expressions are evaluated after the evaluation of any other Expression, even if it returns a `Symbol` or a number. The result of evaluating a `MakeBoxes[]` expression should be a `BoxElementMixin` object, which then can be converted into a  (Python's) `str` which can be shown in the screen.

In Master, all the MakeBoxes rules are stored as builtin rules. In Master, if we set a MakeBoxes rule in an autoload file, these other definitions are erased, and the evaluation gets broken.

Reply:
> I am not sure to understand your question. A `Symbol` in WL is a literal `Atom`, just like a `String` or an `Integer`. With all of them, you can build `Expression` s, which are just a list of Atoms.

Let me review how the code works now, used to work, look at WMA docs and so on. It may take a little time though.